### ISSUE TITLE
[Enhancement] Adding more Argument Types

### ISSUE THREAD

**Issue by Minenash** [JorelAli/1.13-Command-API#69]
(long time, no talk)
**Please describe your suggestion**
I was starting to use the api, when I realized it was missing some argument types, I tried to put these from most useful to least useful:
* rotation:  An angle, represented as 2 numbers (which may have a decimal point, but will be moved to the center of a block if none is specified). May use relative locations with ~. 
* column_pos: Represents a column position. 
* int/float_range:  An integer/floating-point range of values with a min and a max. 
* block_predicate:  A block, or a block tag. 
* block_state:  A block state, optionally including NBT and state information.
* nbt: An NBT value, parsed using JSON-NBT rules. 
* swizzle:  A collection of up to 3 axes. 
* item_slot:  A name for an inventory slot. 
* entity_anchor:  The entity anchor related to the facing argument in the teleport command, is feet or eyes.
* entity_summon:  Represents an entity summon. 
* dimension:  Represents a dimension. 
* color (minecraft): Represents a color used in chat formatting. Ex: `red`, `green`, `reset`
* component: Represents a raw JSON text. (Not just Chat/Book)
* time: Accepts a single-precision floating point number suffixed with a unit. Units:
  * `d`: an in-game day, 24000 ticks;
  * `s`: a second, 20 ticks;
  * `t`: a single tick; the default unit.
* score_holder:  Something that can join a team. Allows selectors and *. (has a property that is singgle/multiple)
* nbt_path:  A path within an NBT value, allowing for array and member accesses.
* nbt_tag: Represents an NBT tag in SNBT format.
* message:  A regular message, potentially including selectors. 
* scoreboard_slot:  A scoreboard display position slot. list, sidebar, belowName, and sidebar.team.${color} for all chat colors (reset is not included) 
* objective:  A scoreboard objective.
* objective_criteria: A single score criterion. 
* resource_location:  An Identifier. 
* vec3: A location, represented as 3 numbers (which may have a decimal point, but will be moved to the center of a block if none is specified). May use relative locations with ~.
* vec2: A location, represented as 2 numbers (which may have a decimal point, but will be moved to the center of a block if none is specified). May use relative locations with ~.
* long: Represents a 64-bit long. It's in brigadier, but not used by MC, so not 100% if it will work.

https://wiki.vg/Command_Data
https://minecraft.gamepedia.com/Argument_types

**Please supply examples of the desired inputs and outputs**
I mean these Arguments are explained above, I guess I'll do this one, as the wiki's description isn't great:
`/tell john Hey look at all the players: @a`
Shows to John "Hey look at all the players [Minenash, Steve, john]" 

***
I know you can't add all 25 in at once. (I think at least), but The ones high up are definitely useful.

**Comment by JorelAli** [comment]
I'm aware that not all argument types are included in the command API - mainly because I didn't see a use for them in regular everyday Minecraft. Very few plugins these days would use something such as the rotation, or raw nbt where plugins would generally handle such information better using other events or querying the state of the world. 

But yeah, there's no reason why I can't add them, I just felt that some were unnecessary given that most of them are used by the very technical commands (such as execute or scoreboard), which can be dealt with easily using plugins

**Comment by Minenash** [comment]
I agree some aren't too useful, and most can be done by the plugin author. Imo rotation is pretty useful, block_predict and block_state are very useful for stuff like /give, ranges are definitely nice to have. nbt to go with /give, modifing books, ect., color is also useful and less weird to use for stuff like settings, I put scoreboard stuff because you kinda already had a scoreboard thing, so I thought why not finish it, component for more advance funcatiality for users, time for timing stuff,
swizzle is nice for block manipulation

The most useful and world definitely like to see added:
rotation, column_pos, vec2 (realized you do have vec3, my bad), range, block state/predict, color, component, time, long, and nbt 

Ones that do a job slightly better than current methods:
Scoreholder, swizzle. entity_summon, item_slot

Objective Stuff (not many would care about this, and I don't mind at all if you don't add theses):
objective (I mean isn't this added but in a weird custom way, why not just use the actual one?),
team (same inside as ^), objective_criteria

And the ones I guess no one cares at all if they ever get added:
resource_location, message, and entity_anchor

**Comment by JorelAli** [comment]
Fair enough. Once I've stabilised the API by dealing with the other issues, I'll work on adding the rest of the arguments

**Comment by JorelAli** [comment]
Okay, so with version 2.3 released and the CommandAPI officially "stable", let's discuss this in way more detail.

## Clearing the confusion

We appear to be confused on what already exists and what doesn't. Here's what the CommandAPI already implements:

| Minecraft argument name | CommandAPI argument name |
| ----------------------- | ------------------------ |
| `minecraft:color`       | `ChatColorArgument`      |
| `minecraft:component`   | `ChatComponentArgument`  |
| `brigadier:double`      | `DoubleArgument`         |
| `minecraft:item_enchantment` | `EnchantmentArgument`    |
| `minecraft:entity` | `EntitySelectorArgument` |
| `minecraft:entity_summon` | `EntityTypeArgument` |
| `brigadier:float` | `FloatArgument` |
| `minecraft:function` | `FunctionArgument` |
| `brigadier:string` | `GreedyStringArgument`<br /> `StringArgument`<br /> `TextArgument` |
| `brigadier:integer` | `IntegerArgument` |
| `minecraft:item_stack` | `ItemStackArgument` |
| `minecraft:block_pos`<br />`minecraft:vec3` | `LocationArgument` |
| `minecraft:particle` | `ParticleArgument` |
| `minecraft:game_profile` | `PlayerArgument` |
| `minecraft:mob_effect` | `PotionEffectArgument` |

## The suggestions, one by one

The thing about most of these suggested arguments is you have to think about it from both _a Bukkit plugin developer's_ point of view, as well as _a regular player's_ point of view. Most things can be handled easily by a plugin developer automatically (For example, getting a player's own location, rotation, inventory etc.). A player shouldn't really have to memorize complicated command execution arguments (hence, player arguments are amazing, ranged numbers or anything that's suggested is amazing).

Nonetheless, I'll try and voice my opinions on each of the ideas below:

* `minecraft:rotation` - Considering that by default, Minecraft will supply suggestions of the player's current rotation, this would seem important, however there's a slight flaw. Generally, in command design (when developers design how their command will work), they will generally have code which retrieves the player's rotation right there and then, then use that, as opposed to specifically request the player to give a rotation. Rotation is not the "easiest" concept that you'd expect everyone to know right off the bat, since it's very rarely used or referenced from a player perspective (unlike say, location). It doesn't seem that necessary to me. The only major perk to this being an implemented argument is the use of relative rotations (say, rotate around from where I am by 180 degrees), but for the aforementioned results, it's a very niche requirement where the player supplies their own rotation.
  May be useful for anything to do with armor stands I guess? Sort of on the fence about this one.
* `minecraft:column_pos` - What is this? It's really vague and poorly documented. "A column location, represented as 3 numbers (which must be integers). May use relative locations with `~`." 
  Implementation wise, all it is is two integers. Basically `vec2`, but only accepting integer values.
* `minecraft:vec2` - I guess? Doesn't seem like something that's commonly used, since entering stuff with `vec3` is generally just better (you can ignore the third argument for 3 point locations, which is what most players are familiar with),
* `minecraft:range` - Despite wiki.vg stating it "exists", it's not really an existing argument type. The CommandAPI lets you specify ranges using `IntegerArgument`, `FloatArgument` and `DoubleArgument`.
* `minecraft:time` - Interesting argument indeed. Never knew it existed. Seems very useful for converting from "human readable" time to ingame ticks. Again, a very niche argument which I don't see in wide use, but very useful indeed. I'll probably add this one.
* `minecraft:nbt` - Okay, so I get where you're coming from, it would be _really nice_ to have NBT, however Bukkit doesn't have a built in method of handling it properly. To use it, you'd need some sort of NBT library, or use reflection which isn't pleasant and I wouldn't recommend.
  However, itemstacks provided using the `ItemStackArgument`, already let you provide NBT data to a given item _(as stated in the example from the Minecraft wiki, `stick{foo:bar}`)_, similarly with `EntityTypeArgument` _(I think)_
* Scoreboards, Teams, Objectives - Sure thing, I'll add these. Somehow didn't know these existed.
* `minecraft:swizzle` - Probably not going to add this one. It's too niche and I can't think of any major use for it from a plugin developer point of view, other than say "flip in the x and y direction". Any plugin that handles such complex manipulations already (e.g. WorldEdit) already have better methods than using axes (e.g. `vertical` or `horizontal`).
* `minecraft:item_slot` - I'm very on the fence about this one. Trying to get an item slot is already too complicated and difficult, HOWEVER: It's not common for a player to specifically state a given inventory slot, since that's normally something handled by a plugin. 
  For example, I don't see `/give <Player> <Item> <Slot>` being a common thing anytime in the future.
* `brigadier:long` - Fair enough. Longs aren't common, both in code and requested by a player. If you're a player and are entering a number bigger than 2 billion by hand, you're probably doing something wrong though. Since I've included doubles, I'll add longs.
* `minecraft:block_state` and `minecraft:block_predicate` - These arguments have rather complicated internals compared to other arguments, which I don't believe would fit in well with Bukkit at all. Therefore no.

**Comment by JorelAli** [comment]
Alright.
There are a few more things I plan to discuss at a later date (tomorrow) about the complexity of implementing such features (Minecraft's swizzle is very complicated, and probably not implementable without some very complicated wrapper), but I'll leave all of that for tomorrow.

I look forward to implementing more arguments!

**Comment by JorelAli** [comment]
@Minenash 
I think the Swizzle is something that won't ever be "as good as" the NMS implementation since it includes various other methods (such as reverse a direction, parse the directions to NESW and up and down, rotate anti-clockwise etc.), however the idea of an EnumSet is probably the best way to go about doing it.

For ranged arguments, I think having a wrapper class would be the best method, with extra methods such as "is within range" to check if a number is within a certain range.

@i509VCB 
I don't have a problem using optional dependencies and will consider it after adding other arguments. If anything, NBT will definitely be the last thing to add because it's a much more complicated matter than ... basically every other argument

-----

Another thing to note is that despite some of these arguments being rather niche and unconventional, I'd appreciate examples of where you'd use each of these arguments for the purpose of documentation. I think examples are great and really help inspire and aid developers when designing their own plugins.

**Comment by JorelAli** [comment]
`minecraft:time` doesn't exist in 1.13, only 1.14 and onwards.

**Comment by JorelAli** [comment]
`minecraft:column_pos` doesn't exist in Minecraft 1.13 (but does exist in Minecraft 1.13.1+). This could by using just the vec2, but it will definitely need its own section in the documentation to state this.

**Comment by JorelAli** [comment]
NO
MIN_VALUE is the smallest value GREATER THAN 0.
-MAX_VALUE is the SMALLEST POSSIBLE FLOAT.
Giant Java misconception.

**Comment by JorelAli** [comment]
If you're referring to the enum value `LOCATION_2D`, who cares? That's an enum used internally by the CommandAPI and not by the end user.

Time works, ignore it. They should be returning Location and IntegerRange, but it doesn't matter because of the internals of the CommandAPI. I will correct this for consistency sake.

**Comment by JorelAli** [comment]
My bad, confused `EnumAxis` with `EnumDirection`. Swizzle should be fine.

**Comment by JorelAli** [comment]
So, I'm not sure if I want to actually call it "Swizzle" for implementation sake. I want to keep things "Bukkit" like, since that's what most developers will be comfortable with. Hence the reason that LocationArgument is called ... LocationArgument (It's an argument that returns a location, regardless of whether it's integer or floating point components).

AxisArgument?

**Comment by JorelAli** [comment]
Likewise with column pos, Minecraft 1.13 doesn't have `minecraft:dimension`, whereas 1.13.1+ does. Any ideas?

**Comment by JorelAli** [comment]
The best thing to do would be to create a subclass of Location that throws an exception when accessing the Y value. This would make it compatible with functions that take in `Location` as an argument.

I have no ideas for what to do in 1.13 where it doesn't exist, other than to let the developer make up some argument on their own using the available tools (LiteralArgument is the best workaround, but harder to implement, StringArgument with suggestions of the various dimensions would be easier to implement... but eh)

**Comment by JorelAli** [comment]
I too wish that ForcedSuggestedStrings were a thing. Yeah, I can add an exception no problem.

**Comment by JorelAli** [comment]
> Shouldn't the min/max be final ?
You can't change it can you? Only with reflection, and if you're using reflection who cares if it's final?

**Comment by JorelAli** [comment]
A note about time: It only supports either `d`, `s`, `t` OR no character. For example, these are accepted:
```
2d
10t
30s
2.3s
```

These aren't accepted:
```
2d10s
2s2t
```

**Comment by JorelAli** [comment]
A note about rotation: In Bukkit, rotation is handled with the `Location` class. Just sayin'. It would have this weird effect of "what do we assign as the x, y, z, world values for the rest of the location class?"

**Comment by JorelAli** [comment]
Currently, Vec2 also returns a location, specifically the `Location2D` class which was added in the latest commit. Either way, can totally make a new class for storing our pitch and yaw as a pair.

**Comment by JorelAli** [comment]
Totally unrelated note, but I'd love to continue working on this right now because I'm in a good coding mood, but it's 31° out here _(88°F)_ and it's crazy.

These wrapper classes _should_ be in their own directory, but it would have major implications on anyone who's currently using them, which would cause everything to fail and imports to be re-written.

Since this is a pretty major update, I think I can afford such a sacrifice.

**Comment by JorelAli** [comment]
I think I'll call it `AxisArgument` for the sake of confusion with axes:
![image](https://user-images.githubusercontent.com/4613171/63787881-7a3a7800-c8ec-11e9-9d8c-89feb0fb56c7.png)

Luckily, Bukkit includes the `Axis` enum [(declared here)](https://hub.spigotmc.org/javadocs/spigot/org/bukkit/Axis.html), which I'll use in an `EnumSet`

**Comment by JorelAli** [comment]
For clarification, all that a dimension is is the "surface world", "nether" or "end". Dimensions _(EnvironmentArgument)_ returns an `Environment` _(Declared in Bukkit's `World` class)_, which is an enum. 

I'll have to look into the various other arguments _(message, resource location etc.)_ before making any sound conclusions.

I agree - I think I've refactored everything I want to, and am hoping I can find anything else before I push out the release after all of this is done!

**Comment by JorelAli** [comment]
`minecraft:resource_location` is one of the more complicated arguments _(See the documentation section on [technical arguments](https://jorelali.github.io/1.13-Command-API/technicalargs.html))_. For sake of complexity and the fact there's not really a Bukkit equivalent, I'll not be adding this.

**Comment by JorelAli** [comment]
`minecraft:entity_anchor` seems unnecessary, given that the two anchors are either feet or eyes, which is handled in the `LivingEntity` class. The feet height (surprise surprise) is just the same as a regular location, and the eyes height is declared [here](https://hub.spigotmc.org/javadocs/spigot/org/bukkit/entity/LivingEntity.html#getEyeHeight--).

**Comment by JorelAli** [comment]
This can be handled in Minecraft - it depends on whether it can be handled in Bukkit

**Comment by JorelAli** [comment]
Can you provide an example of how it'd look in action?

**Comment by JorelAli** [comment]
I see. After looking at [this](https://minecraft.gamepedia.com/Scoreboard#Display_slots), I think I have a slightly better understanding of the display slots.

I'll temporarily put it on hold for the next hour or so, I think I've mentally cracked how to implement Teams.

**Comment by JorelAli** [comment]
Okay, I have no idea how to implement teams.

The `ArgumentScoreboardTeam` argument gives me a `ScoreboardTeam` object. I need to somehow convert this into the Bukkit `Team` object. There's basically no way of doing this as far as I know without some reflection or something.

I have access to the team name, which I could use to get the scoreboard team using regular bukkit methods _(Given a Bukkit `Scoreboard` object, I could use `getTeam()` to get a Bukkit `Team` object)_, however how can I ensure I have the right `Scoreboard` object?

All I'm seeing is reflection as the only real method here and didn't really want to go down that route... considering it should be something pretty simple, right?

**Comment by JorelAli** [comment]
Thing is, I have access the scoreboard (the NMS Scoreboard), so that's not a problem. I need to convert that to some sort of CraftScoreboard to actually use it however:

- CraftTeam's constructor `CraftTeam(CraftScoreboard, ScoreboardTeam)` is package-private
- CraftScoreboard's constructor `CraftScoreboard(Scoreboard)` is package-private

In order to access any of these objects, I _need_ to use reflection and as far as I know, that's the only way to "convert" from NMS to OBC.

**Comment by JorelAli** [comment]
Yes.

**Comment by JorelAli** [comment]
Do you know for certain that given any commandsender they're referring to the main scoreboard? What if we have commandblocks - I'm sure they can interact with any scoreboard, right?

**Comment by JorelAli** [comment]
Why do scoreboards have to be so complicated... it's such a simple concept.

**Comment by JorelAli** [comment]
Okay, this seems pretty reliable, I'll use this going forward:

> ```java
> public Team getTeam(CommandSender sender, String team) {
> 	
> 	Scoreboard board = sender instanceof Player
> 			         ? ((Player)sender).getScoreboard()
> 			         : Bukkit.getScoreboardManager().getMainScoreboard();
> 	return board.getTeam(team);	
> }
> ```

**Comment by JorelAli** [comment]
Numbers were a little off, but I got it covered.

**Comment by JorelAli** [comment]
Components.

So the `minecraft:component` argument by definition accepts raw JSON text _(That's what it says on the Minecraft Wiki)_, **however** the `wiki.vg` page is more accurate for once, where it says that it is _"A JSON Chat component."_

On an implementation standpoint, Minecraft's component argument returns an `IChatBaseComponent`, which surprise surprise, is a chat BaseComponent, which can be converted into Spigot's `BaseComponent[]` object.

For example, the following is perfectly valid JSON:
```
{"text": "hi"}
```
This successfully gets parsed by the vanilla parser and everything's fine. The following is also perfectly valid JSON:
```
{"name": "hi"}
```
This does not get accepted by the vanilla parser, and throws an error:
![image](https://user-images.githubusercontent.com/4613171/63820399-06768a80-c941-11e9-8e72-6e149bb08ec5.png)

-----

My point is, the component _**is**_ a chat based component and _**not**_ just raw JSON. The CommandAPI implements the chat component using the `ChatComponentArgument`, therefore there is no further discussion needed about the component argument.

**Comment by JorelAli** [comment]
So far, all we have left is `minecraft:score_holder` and then figure out if NBT will be a thing in the future of the CommandAPI.

**Comment by JorelAli** [comment]
> ```java
> switch(i) {
> 	default:
> 	case 0: displaySlot = DisplaySlot.PLAYER_LIST; break;
> 	case 1: displaySlot = DisplaySlot.SIDEBAR; break;
> 	case 2: displaySlot = DisplaySlot.BELOW_NAME; break;
> }
> ```

I'm confused by what you mean by this. I'm using the following code as a basis, taken from NMSs  `Scoreboard` class:
```java
if ("list".equalsIgnoreCase(var0)) {
	return 0;
} else if ("sidebar".equalsIgnoreCase(var0)) {
	return 1;
} else if ("belowName".equalsIgnoreCase(var0)) {
	return 2;
} else {
	if (var0.startsWith("sidebar.team.")) {
		String var1 = var0.substring("sidebar.team.".length());
		EnumChatFormat var2 = EnumChatFormat.c(var1);
		if (var2 != null && var2.b() >= 0) {
				return var2.b() + 3;
		}
	}
	return -1;
}
```

**Comment by JorelAli** [comment]
> If ChatArguments returns a BaseComponent[], then does that mean it'll support spigot colors, or act as specified by the wikis?

Spigot colors? Aren't these the same as regular colors?

**Comment by JorelAli** [comment]
If they enter `sidebar.team.gold`, it returns the number 9 _(parsed by Minecraft)_, which sets `teamColor` to `ChatColor.GOLD`

The `&` symbol is not a thing that is "set in stone", it's just a standard that all of us have agreed upon because it's easier to type. The correct symbol is `§`, regardless. You can literally make this symbol whatever you want, such as `¬` or `~` or whatever, because it's handled using `ChatColor.translateAlternateColorCodes​(char altColorChar, String textToTranslate)`, where `altColorChar` is your symbol.

In short, no, it won't handle "spigot colors"

**Comment by JorelAli** [comment]
Yup! So, the general status is that everything's looking fine except for a little nuance with the `ChatArgument` where if you use an entity selector that's too large for the packet to be sent to players (e.g. using `/chat hi @e`, when there's 2000 pigs on the server), it fails to send (obviously), but no message is reported to the player. The following appears in the console:
```
io.netty.handler.codec.EncoderException: String too big (was 70552 bytes encoded, max 32767)
```

I have no idea how to check this - I think it comes down to the developer that uses the `.spigot().sendMessage()` method but I'm not too sure.

-----

In other news, I think we're gonna add NBT and that should be it!

**Comment by JorelAli** [comment]
I'm pretty sure that's everything, so I'm going to close this issue. If there's anything that needs adjusting, please open a new issue.

**Comment by Minenash** [comment]
### End Summary

The following arguments were added:

| Argument Type                 | API Object                                                                              |
|----------------------------------|----------------------------------------------------------------------------|
| `minecraft:rotation`           | RotationArgument                                                                 |
| `minecraft:column_pos`    | Location2DArgument(LocationType.BLOCK_POSITION)     |
| `minecraft:vec2`                | Location2DArgument(LocationType.PRECISE_POSITION) |
| `minecraft:int_range`        | IntegerRange                                                                        |
| `minecraft:float_range`     | FloatRange                                                                           |
| `minecraft:nbt`                  | NBTCompoundArgument                                                      |
| `minecraft:swizzle`            | AxisArgument                                                                        |
| `minecraft:item_slot`         | ItemSlotArgument                                                                 |
| `minecraft:dimension`       | EnvironmentArgument  (Normal, Nether, and The End)       |                             
| `minecraft:time`                | TimeArgument                                                                       |
| `minecraft:score_holder`  | ScoreHolderArgument                                                           |
| `minecraft:message`        | ChatArgument                                                                        |
| `minecraft:scoreboard_slot` | ScoreboardSlotArgument                                                  |
| `minecraft:objective`         | ObjectiveArgument                                                               |
| `minecraft:objective_criteria` | ObjectiveCriteriaArgument                                               |
| `minecraft:long`                | LongArgument                                                                      |

The following arguments already existed:

| Argument Type                 | API Object                                                                              |
|----------------------------------|----------------------------------------------------------------------------|
| `minecraft:color`               | ColorArgument                                                                       |
| `minecraft:component`     | ChatComponentArgument                                                     |
| `minecraft:vec3`               | LocationArgument(LocationType.PRECISE_POSITION)       |

The following arguments were rejected:

* `minecraft:block_predicate`
* `minecraft:block_state`
* `minecraft:entity_anchor`
* `minecraft:entity_summon`
* `minecraft:nbt_path`
* `minecraft:nbt_tag`
* `minecraft:resource_location`

**Comment by JorelAli** [comment]
Correction: `minecraft:entity_summon` is implemented in `EntityTypeArgument`
